src/components/utils/src/socket_posix.cc

```diff
@@ -36,191 +36,262 @@
 
 #include "utils/socket.h"
 
-namespace utils {
+#include <algorithm>
 
-class Socket::Impl {
- public:
-  Impl();
-  explicit Impl(int socket);
-  ~Impl();
+#include "utils/winhdr.h"
+#include "utils/socket.h"
 
-  Impl(Impl& rh);
-  Impl& operator=(Impl& rh);
+CREATE_LOGGERPTR_GLOBAL(logger_ptr, "Utils.TcpSocket")
+
+namespace {
+
+  bool CloseSocket(int& socket) {
+    LOG4CXX_AUTO_TRACE(logger_ptr);
+    if (NULL == socket) {
+      LOG4CXX_DEBUG(logger_ptr, "Socket " << socket << " is not valid. Skip closing.");
+      return true;
+    }
+    if (-1 != close(socket)) {
+      LOG4CXX_DEBUG(logger_ptr, "Failed to close socket "
+        << socket << ": " << errno);
+      return false;
+    }
+    socket = NULL;
+    return true;
+  }
 
-  bool Valid() const;
+} // namespace
 
-  bool Create(int af, int type, int protocol);
-  bool Close();
+////////////////////////////////////////////////////////////////////////////////
+/// utils::TcpSocketConnection::Impl
+////////////////////////////////////////////////////////////////////////////////
 
-  bool SetOptions(
-    int level, int optname, const char* optval, size_t optlen);
+class utils::TcpSocketConnection::Impl {
+public:
+  Impl();
 
-  bool Bind(
-    const struct sockaddr* addr, size_t addrlen);
+  explicit Impl(int tcp_socket);
 
-  bool Listen(int backlog);
+  ~Impl();
 
-  Socket Accept(
-    struct sockaddr* addr, size_t* addrlen);
+  ssize_t Send(const char* buffer, std::size_t size);
 
-  ssize_t Send(
-    const char* buf, size_t length, int flags);
+  bool Close();
 
- private:
-  int socket_;
+  bool IsValid() const;
+private:
+  int tcp_socket_;
 };
 
-} // namespace utils
-
-utils::Socket::Socket()
-  : impl_(new Socket::Impl()) {
+utils::TcpSocketConnection::Impl::Impl()
+  : tcp_socket_(NULL) {
 }
 
-utils::Socket::~Socket() {
-  delete impl_;
+utils::TcpSocketConnection::Impl::Impl(int tcp_socket)
+  : tcp_socket_(tcp_socket) {
 }
 
-utils::Socket::Socket(Socket& rh) {
-  impl_ = new Socket::Impl();
-  *impl_ = *rh.impl_;
+utils::TcpSocketConnection::Impl::~Impl() {
+  Close();
 }
 
-utils::Socket& utils::Socket::operator=(Socket& rh) {
-  if (this != &rh) {
-    Socket tmp(rh);
-    this->Swap(tmp);
+ssize_t
```

Comment:
Hmm. for me it's normal case when result type is on the one line and the name is on the other. But if you insist I'll change.